### ISSUE TITLE
Jet stuck on starting Job

### ISSUE THREAD

**Issue by TomaszGaweda** [hazelcast/hazelcast-jet#1339]
I'm starting many jobs at once (~30, some are triggered slightly before other). However, when my Hazelcast Jet job count hits 26 (magic number?), all processing got stuck.

In the threadumps I see following info:
```
"hz._hzInstance_1_jet.cached.thread-1" #37 prio=5 os_prio=0 cpu=1093.29ms elapsed=393.62s tid=0x00007f95dc007000 nid=0x6bfc in Object.wait()  [0x00007f95e6af4000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
    at java.lang.Object.wait(java.base@11.0.2/Native Method)
    - waiting on <no object reference available>
    at com.hazelcast.spi.impl.AbstractCompletableFuture.get(AbstractCompletableFuture.java:229)
    - waiting to re-lock in wait() <0x00000007864b7040> (a com.hazelcast.internal.util.SimpleCompletableFuture)
    at com.hazelcast.spi.impl.AbstractCompletableFuture.get(AbstractCompletableFuture.java:191)
    at com.hazelcast.spi.impl.operationservice.impl.InvokeOnPartitions.invoke(InvokeOnPartitions.java:88)
    at com.hazelcast.spi.impl.operationservice.impl.OperationServiceImpl.invokeOnAllPartitions(OperationServiceImpl.java:385)
    at com.hazelcast.map.impl.proxy.MapProxySupport.clearInternal(MapProxySupport.java:1016)
    at com.hazelcast.map.impl.proxy.MapProxyImpl.clearInternal(MapProxyImpl.java:109)
    at com.hazelcast.map.impl.proxy.MapProxyImpl.clear(MapProxyImpl.java:698)
    at com.hazelcast.jet.impl.JobRepository.clearSnapshotData(JobRepository.java:464)
    at com.hazelcast.jet.impl.MasterJobContext.tryStartJob(MasterJobContext.java:233)
    at com.hazelcast.jet.impl.JobCoordinationService.tryStartJob(JobCoordinationService.java:776)
    at com.hazelcast.jet.impl.JobCoordinationService.lambda$submitJob$0(JobCoordinationService.java:200)
    at com.hazelcast.jet.impl.JobCoordinationService$$Lambda$634/0x00000008009ce840.run(Unknown Source)
```

and also:

```
"hz._hzInstance_1_jet.async.thread-2" #81 prio=5 os_prio=0 cpu=0.00ms elapsed=661.98s tid=0x0000025bb23ef000 nid=0x43bc in Object.wait()  [0x0000005d492fe000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
    at java.lang.Object.wait(java.base@11/Native Method)
    - waiting on <no object reference available>
    at com.hazelcast.spi.impl.AbstractCompletableFuture.get(AbstractCompletableFuture.java:229)
    - waiting to re-lock in wait() <0x0000000725600100> (a com.hazelcast.internal.util.SimpleCompletableFuture)
    at com.hazelcast.spi.impl.AbstractCompletableFuture.get(AbstractCompletableFuture.java:191)
    at com.hazelcast.spi.impl.operationservice.impl.InvokeOnPartitions.invoke(InvokeOnPartitions.java:88)
    at com.hazelcast.spi.impl.operationservice.impl.OperationServiceImpl.invokeOnAllPartitions(OperationServiceImpl.java:385)
    at com.hazelcast.map.impl.proxy.MapProxySupport.removeAllInternal(MapProxySupport.java:619)
    at com.hazelcast.map.impl.proxy.MapProxyImpl.removeAll(MapProxyImpl.java:285)
    at com.hazelcast.jet.impl.JobRepository.deleteJob(JobRepository.java:332)
    at com.hazelcast.jet.impl.JobRepository.completeJob(JobRepository.java:316)
    at com.hazelcast.jet.impl.JobCoordinationService.completeJob(JobCoordinationService.java:576)
    at com.hazelcast.jet.impl.MasterJobContext.lambda$finalizeJob$13(MasterJobContext.java:620)
    at com.hazelcast.jet.impl.MasterJobContext$$Lambda$783/0x0000000800b26840.run(Unknown Source)
    at com.hazelcast.jet.impl.MasterJobContext.finalizeJob(MasterJobContext.java:632)
    at com.hazelcast.jet.impl.MasterJobContext.onCompleteExecutionCompleted(MasterJobContext.java:564)
    at com.hazelcast.jet.impl.MasterJobContext.lambda$invokeCompleteExecution$6(MasterJobContext.java:544)
    at com.hazelcast.jet.impl.MasterJobContext$$Lambda$779/0x0000000800b27840.accept(Unknown Source)
    at com.hazelcast.jet.impl.MasterContext.lambda$invokeOnParticipants$0(MasterContext.java:242)
    at com.hazelcast.jet.impl.MasterContext$$Lambda$726/0x0000000800a1c040.accept(Unknown Source)
    at com.hazelcast.jet.impl.util.Util$2.onResponse(Util.java:172)
    at com.hazelcast.spi.impl.AbstractInvocationFuture$1.run(AbstractInvocationFuture.java:256)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@11/ThreadPoolExecutor.java:1128)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@11/ThreadPoolExecutor.java:628)
    at java.lang.Thread.run(java.base@11/Thread.java:834)
    at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64)
    at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80)
```

My setup: 
- Java 11 
- Hazelcast 3.12 Snapshot 
- Hazelcast Jet 3.0 Snapshot (I can't revert to previous version, it will break my logic; I need n:m joins which will be added in 3.1) 
- CPU cores: 4 
- RAM: 7 GB 
- Jet mode: server, connects to other cluster as a client to insert final data

Simple repro:
```
import com.hazelcast.core.IMap;
import com.hazelcast.jet.Jet;
import com.hazelcast.jet.JetInstance;
import com.hazelcast.jet.core.JobStatus;
import com.hazelcast.jet.pipeline.Pipeline;
import com.hazelcast.jet.pipeline.Sources;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.stream.Collectors;

import static com.hazelcast.jet.aggregate.AggregateOperations.summingDouble;
import static com.hazelcast.jet.pipeline.Sinks.logger;
import static java.math.RoundingMode.HALF_UP;
import static java.util.stream.Collectors.groupingBy;

public class JobStuckingReproducer {

    private static final int NUMBER_OF_JOBS = 30;
    private static volatile boolean shouldRun = true;

    public static void main(String[] args) throws ExecutionException, InterruptedException {
        JetInstance jet = Jet.newJetInstance();
        List<IMap<Integer, Integer>> maps = new ArrayList<>();
        for (int i = 0; i < NUMBER_OF_JOBS; i++) {
            var map = jet.<Integer, Integer>getMap("map" + i);
            for (int j = 0; j < 10_000; j++) {
                map.put(j, j);
            }
            maps.add(map);
        }

        List<CompletableFuture<Void>> futures = new ArrayList<>();
        for (int i = 0; i < NUMBER_OF_JOBS; i++) {
            var map = maps.get(i);
            Pipeline p = Pipeline.create();
            p.drawFrom(Sources.map(map))
                    .map(e -> {
                        var bd = new BigDecimal(e.getValue());

                        return bd.pow(11).subtract(bd.divide(new BigDecimal(19), HALF_UP)); // simulate any CPU usage
                    })
                    .groupingKey(e -> e.setScale(0, HALF_UP).intValue() % 10)
                    .aggregate(summingDouble(e -> e.doubleValue()))
                    .drainTo(logger());

            var job = jet.newJob(p);
            futures.add(job.getFuture());
        }

        new Thread(() -> {
            while (shouldRun) {
                try {
                    Thread.sleep(2000);

                    var jobByStatus = jet.getJobs().stream().collect(groupingBy(e -> e.getStatus(), Collectors.counting()));
                    for (JobStatus status : jobByStatus.keySet()) {
                        System.out.println("\t" + status + " -> " + jobByStatus.get(status));
                    }
                    System.out.println("in total: " + jet.getJobs().size());
                } catch (InterruptedException e) {
                    throw new RuntimeException(e);
                }
            }
        }, "jet-jobs-printer").start();

        var combined = CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));

        combined.get();
        shouldRun = false;
        jet.shutdown();
    }
}
```

**Comment by TomaszGaweda** [comment]
@cangencer Hi, this is a bug you've commented on on StackOverflow. I did manage to create a simplified repro

**Comment by cangencer** [comment]
Thanks, we reproduced it and are working on a fix.

**Comment by cangencer** [comment]
@TomaszGaweda this should be fixed now by #1340 on the latest snapshot
